### PR TITLE
Enable trimming and fix errors in multiple projects

### LINKED ISSUE #8925: IL2026: Members attributed with RequiresUnreferencedCode may break when trimming
**Describe the bug**
This issue is for track possible trimming issue [IL2026](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-warnings/il2026?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(IL2026)%26rd%3Dtrue)

### List of warnings:

|Severity|Code|Description|Project|File|Line|Suppression State
|-|-|-|-|-|-|-|
Warning|IL2026|Using member 'System.ComponentModel.TypeDescriptor.GetConverter(System.Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Animation\AnimatorKeyFrame.cs|71|N/A
Warning|IL2026|Using member 'System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(System.RuntimeTypeHandle)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Trimmer can't guarantee existence of class constructor.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\AvaloniaPropertyRegistry.cs|60|N/A
Warning|IL2026|Using member 'System.ComponentModel.DataAnnotations.ValidationContext.ValidationContext(object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. The Type of instance cannot be statically discovered.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Data\Core\Plugins\DataAnnotationsValidationPlugin.cs|41|N/A
Warning|IL2026|Using member 'System.ComponentModel.DataAnnotations.Validator.TryValidateProperty(object?, System.ComponentModel.DataAnnotations.ValidationContext, System.Collections.Generic.ICollection<System.ComponentModel.DataAnnotations.ValidationResult>?)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. The Type of validationContext.ObjectType cannot be statically discovered.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Data\Core\Plugins\DataAnnotationsValidationPlugin.cs|53|N/A
Warning|IL2026|Using member 'System.Type.GetType(string)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. The type might be removed.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Platform\StandardRuntimePlatform.cs|38|N/A
Warning|IL2026|Using member 'System.ComponentModel.TypeDescriptor.GetConverter(System.Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Styling\PropertyEqualsSelector.cs|110|N/A
Warning|IL2026|Using member 'System.Runtime.Serialization.XmlObjectSerializer.WriteObject(System.IO.Stream, object?)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Data Contract Serialization and Deserialization might require types that cannot be statically analyzed. Make sure all of the required types are preserved.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Utilities\AvaloniaResourcesIndex.cs|41|N/A
Warning|IL2026|Using member 'System.ComponentModel.TypeDescriptor.GetConverter(System.Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Utilities\TypeUtilities.cs|211|N/A
Warning|IL2026|Using member 'System.ComponentModel.TypeDescriptor.GetConverter(System.Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.|Avalonia.Base (net6.0)|.\src\Avalonia.Base\Utilities\TypeUtilities.cs|219|N/A

**To Reproduce**

1. Rebuild Avalonia solution
2. See warnings

**Expected behavior**
no warnings

**Screenshots**

**Desktop (please complete the following information):**

- OS:
- Version da43d2a5fa7774cd88053affec7434af411f39d0

**Additional context**

### LINKED ISSUE #8944: Enable Analyzers
Add NetAnalyzers  for monitor performance issue

- [x] [CA1304](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1304): Specify CultureInfo #9588
- [x] [CA1802](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1802): Use literals where appropriate #9162
- [x]  ~~[CA1805](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1805): Do not initialize unnecessarily~~ see [comment](https://github.com/AvaloniaUI/Avalonia/pull/9163#issuecomment-1276831556)
- [ ] [CA1806](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806): Do not ignore method results 
- [ ] [CA1810](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1810): Initialize reference type static fields inline
- [ ] [CA1812](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1812): Avoid uninstantiated internal classes 
- [ ] [CA1813](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1813): Avoid unsealed attributes
- [ ] [CA1815](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1815): Override equals and operator equals on value types
- [ ] [CA1819](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819): Properties should not return arrays
- [x] [CA1820](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1820): Test for empty strings using string length #9188 
- [x] [CA1821](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1821): Remove empty Finalizers
- [x] [CA1822](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822): Mark members as static #9189
- [ ] [CA1823](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1823): Avoid unused private fields
- [x] [CA1825](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825): Avoid zero-length array allocations #9166
- [x] [CA1826](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1826): Do not use Enumerable methods on indexable collections #9593
- [x] [CA1827](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1827): Do not use Count() or LongCount() when Any() can be used #9593
- [x] [CA1828](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1828): Do not use CountAsync() or LongCountAsync() when AnyAsync() can be used #9593
- [x] [CA1829](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1829): Use Length/Count property instead of Count() when available #9593
- [ ] [CA1835](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1835): Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync'
- [ ] [CA1838](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1838): Avoid 'StringBuilder' parameters for P/Invokes
- [ ] [CA1844](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1844): Provide memory-based overrides of async methods when subclassing 'Stream'
- [ ] [CA1846](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1846): Prefer 'AsSpan' over 'Substring'
- [x] [CA1847](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1847): Use char literal for a single character lookup #9190
- [ ] [CA1848](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1848): Use the LoggerMessage delegates
- [ ] [CA1849](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1849): Call async methods when in an async method

### PR DESCRIPTION
**Pull request by maxkatz6**
Idea is to annotate every member that uses dynamic code/reflection with special attributes. This will work as hints for the linker. Also, as bonus, during release build developers will see warnings/errors, if they use any potentially dangerous type in code (including XAML), for instance - ReflectionBinding.
In some places I had to put "suppress" attributes, mostly related to the implicit type conversion. I don't see any way to prevent linker from removing "op_Implicit" methods. And I don't think it's reasonable to cover whole binding/valuestore API with these attributes.

Also fixes NativeMenuBar not working with AOT.

## Trimmable assemblies:

- Avalonia.Base
- Avalonia.Controls
- Avalonia.Controls.ColorPicker
- Avalonia.Dialogs
- Avalonia.Headless
- Avalonia.Headless.Vnc
- Avalonia.MicroCom
- Avalonia.Native
- Avalonia.OpenGL
- Avalonia.ReactiveUI
- Avalonia.Remote.Protocol
- Avalonia.Themes.Fluent
- Avalonia.Themes.Simple
- Avalonia.Direct2D1
- Avalonia.Win32 (but Automation PR is disabled when COM is not available)
- Avalonia.Skia
- Avalonia.Desktop
- Avalonia.LinuxFramebuffer
- Avalonia.X11
- Avalonia.iOS
- Avalonia.Android
- Avalonia.Browser
- Avalonia.Browser.Blazor

## Untrimmable assemblies:

- all tests
- all samples
- Avalonia.Controls.DataGrid
- Avalonia.Diagnostics
- Avalonia.DesignerSupport (used only for the designer)
- Avalonia.Designer.HostApp (used only for the designer)
- Avalonia.Build.Tasks (used only compiled time)
- Avalonia.FreeDesktop (DBus should be replaced with source generators first)
- Avalonia.Win32.Interop (because of WPF)

## Fixed issues

Fixes #8925
Partially fixes #8944

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026399-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: build/TrimmingEnable.props [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    <!-- Trim warnings -->
+    <WarningsAsErrors>$(WarningsAsErrors);IL2000;IL2001;IL2002;IL2003;IL2004;IL2005;IL2006;IL2007;IL2008;IL2009;IL2010;IL2011;IL2012;IL2013;IL2014;IL2015;IL2016;IL2017;IL2018;IL2019;IL2020;IL2021;IL2022;IL2023;IL2024;IL2025;IL2026;IL2027;IL2028;IL2029;IL2030;IL2031;IL2032;IL2033;IL2034;IL2035;IL2036;IL2037;IL2038;IL2039;IL2040;IL2041;IL2042;IL2043;IL2044;IL2045;IL2046;IL2047;IL2048;IL2049;IL2050;IL2051;IL2052;IL2053;IL2054;IL2055;IL2056;IL2057;IL2058;IL2059;IL2060;IL2061;IL2062;IL2063;IL2064;IL2065;IL2066;IL2067;IL2068;IL2069;IL2070;IL2071;IL2072;IL2073;IL2074;IL2075;IL2076;IL2077;IL2078;IL2079;IL2080;IL2081;IL2082;IL2083;IL2084;IL2085;IL2086;IL2087;IL2088;IL2089;IL2090;IL2091;IL2092;IL2093;IL2094;IL2095;IL2096;IL2097;IL2098;IL2099;IL2100;IL2101;IL2102;IL2103;IL2104;IL2105;IL2106;IL2107;IL2108;IL2109;IL2110;IL2111;IL2112;IL2113;IL2114;IL2115;IL2116;IL2117;IL2118;IL2119;IL2120;IL2121;IL2122;IL2123;IL2124;IL2125;IL2126;IL2127;IL2128;IL2129;IL2130;IL2131;IL2132;IL2133;IL2134;IL2135;IL2136;IL2137;IL2138;IL2139;IL2140;IL2141;IL2142;IL2143;IL2144;IL2145;IL2146;IL2147;IL2148;IL2149;IL2150;IL2151;IL2152;IL2153;IL2154;IL2155;IL2156;IL2157</WarningsAsErrors>
```
Comment (maxkatz6) [inline build/TrimmingEnable.props:12]: For some reason single ILLinkTreatWarningsAsErrors doesn't enable every trimming warning as an error. This works tho...

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027045-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027070-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**